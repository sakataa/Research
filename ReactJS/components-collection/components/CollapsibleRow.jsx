import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

export default class CollapsibleRow extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            isCollapse : !this.props.isExpand
        };
    };

    static propTypes = {
        isExpand: PropTypes.bool,
        isHtml: PropTypes.bool,
        subject: PropTypes.string,
        content: PropTypes.string,
        onExpand: PropTypes.func
    }

    componentWillReceiveProps(nextProps) {
        if( this.props.isExpand !== nextProps.isExpand ) {
            this.setState({ isCollapse : !nextProps.isExpand });
        }
    }

    onClick = () => {
        const newState = !this.state.isCollapse;

        this.setState({ isCollapse : newState });

        if (!newState) {
            this.props.onExpand && this.props.onExpand();
        }
    }

    render() {
        const { subject, content, isHtml } = this.props;

        return <div className={this.state.isCollapse ? "collapse-item collapse" : "collapse-item"}>
                    <div title={subject} className="subject-item"><span>{subject}</span><i className="icon-arrow" onClick={this.onClick}></i></div>
                    {
                        isHtml ? 
                            <div className="content-item" dangerouslySetInnerHTML={{ __html: content }} /> :
                            <div className="content-item">{content}</div>                    
                    }
               </div>
    }
};