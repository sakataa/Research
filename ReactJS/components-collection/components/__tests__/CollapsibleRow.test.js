import React from 'react'
import renderer from 'react-test-renderer'

import CollapsibleRow from '../CollapsibleRow'

describe('CollapsibleRow ', () => {
    describe('render', () => {
        it('isExpand = true - component is expanded', () => { 
            const props = {
                isExpand: false
            };

            const component = renderer.create(<CollapsibleRow {...props} />);

            const result = component.toJSON(); 
        
            expect(result).toMatchSnapshot(); 
        });

        it('isExpand = false - component is collapsed', () => { 
            const props = {
                isExpand: false
            };

            const component = renderer.create(<CollapsibleRow {...props} />);

            const result = component.toJSON(); 
        
            expect(result).toMatchSnapshot(); 
        });

        it('isHtml = false - component render correctly - show raw content', () => { 
            const props = {
                isHtml: false,
                subject: 'Subject',
                content: '<p>Content</p>'
            };

            const component = renderer.create(<CollapsibleRow {...props} />);

            const result = component.toJSON(); 
        
            expect(result).toMatchSnapshot(); 
        });

        it('isHtml = true - component render correctly - show content with Html format', () => { 
            const props = {
                isHtml: true,
                subject: 'Subject',
                content: '<p>Content</p>'
            };

            const component = renderer.create(<CollapsibleRow {...props} />);

            const result = component.toJSON(); 
        
            expect(result).toMatchSnapshot();
        });
    });

    describe('componentWillReceiveProps', () => {
        it('props is not changed - state is not changed', () => { 
            const props = {
                isExpand: false
            };
            let component = new CollapsibleRow(props);
            component.setState = jest.fn();

            const newProps = {
                isExpand: false
            };
            component.componentWillReceiveProps(newProps);

            expect(component.setState).not.toHaveBeenCalled();
        });

        it('props is changed - state is changed', () => { 
            const props = {
                isExpand: false
            };
            let component = new CollapsibleRow(props);
            component.setState = jest.fn();

            const newProps = {
                isExpand: true
            };
            component.componentWillReceiveProps(newProps);

            expect(component.setState).toHaveBeenCalled();
        });
    });

    describe('onClick', () => {
        it('Class name is changed to show/hide content', () => {
            const props = {
                isHtml: false,
                subject: 'Subject',
                content: '<p>Content</p>'
            };

            const component = renderer.create(<CollapsibleRow {...props} />);
            
            let tree = component.toJSON();      
            expect(tree).toMatchSnapshot();

            // expand
            tree.children[0].children[1].props.onClick();           
            tree = component.toJSON();      
            expect(tree).toMatchSnapshot();

            // collapse
            tree.children[0].children[1].props.onClick();
            tree = component.toJSON();      
            expect(tree).toMatchSnapshot();
        });

        it('onExpand function is called when expanding', () => {
            const props = {
                isHtml: false,
                subject: 'Subject',
                content: '<p>Content</p>',
                onExpand: jest.fn(),
                isExpand: true
            };

            const component = renderer.create(<CollapsibleRow {...props} />);
            
            let tree = component.toJSON();      
            expect(tree).toMatchSnapshot();

            // collapse
            tree.children[0].children[1].props.onClick();
            expect(props.onExpand).not.toHaveBeenCalled();

            // expand
            tree.children[0].children[1].props.onClick();
            expect(props.onExpand).toHaveBeenCalled();
        });
    });
});