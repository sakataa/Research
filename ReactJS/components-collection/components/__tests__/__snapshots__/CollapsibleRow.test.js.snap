// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleRow  onClick Class name is changed to show/hide content 1`] = `
<div
  className="collapse-item collapse"
>
  <div
    className="subject-item"
    title="Subject"
  >
    <span>
      Subject
    </span>
    <i
      className="icon-arrow"
      onClick={[Function]}
    />
  </div>
  <div
    className="content-item"
  >
    &lt;p&gt;Content&lt;/p&gt;
  </div>
</div>
`;

exports[`CollapsibleRow  onClick Class name is changed to show/hide content 2`] = `
<div
  className="collapse-item"
>
  <div
    className="subject-item"
    title="Subject"
  >
    <span>
      Subject
    </span>
    <i
      className="icon-arrow"
      onClick={[Function]}
    />
  </div>
  <div
    className="content-item"
  >
    &lt;p&gt;Content&lt;/p&gt;
  </div>
</div>
`;

exports[`CollapsibleRow  onClick Class name is changed to show/hide content 3`] = `
<div
  className="collapse-item collapse"
>
  <div
    className="subject-item"
    title="Subject"
  >
    <span>
      Subject
    </span>
    <i
      className="icon-arrow"
      onClick={[Function]}
    />
  </div>
  <div
    className="content-item"
  >
    &lt;p&gt;Content&lt;/p&gt;
  </div>
</div>
`;

exports[`CollapsibleRow  onClick onExpand function is called when expanding 1`] = `
<div
  className="collapse-item"
>
  <div
    className="subject-item"
    title="Subject"
  >
    <span>
      Subject
    </span>
    <i
      className="icon-arrow"
      onClick={[Function]}
    />
  </div>
  <div
    className="content-item"
  >
    &lt;p&gt;Content&lt;/p&gt;
  </div>
</div>
`;

exports[`CollapsibleRow  render isExpand = false - component is collapsed 1`] = `
<div
  className="collapse-item collapse"
>
  <div
    className="subject-item"
    title={undefined}
  >
    <span />
    <i
      className="icon-arrow"
      onClick={[Function]}
    />
  </div>
  <div
    className="content-item"
  />
</div>
`;

exports[`CollapsibleRow  render isExpand = true - component is expanded 1`] = `
<div
  className="collapse-item collapse"
>
  <div
    className="subject-item"
    title={undefined}
  >
    <span />
    <i
      className="icon-arrow"
      onClick={[Function]}
    />
  </div>
  <div
    className="content-item"
  />
</div>
`;

exports[`CollapsibleRow  render isHtml = false - component render correctly - show raw content 1`] = `
<div
  className="collapse-item collapse"
>
  <div
    className="subject-item"
    title="Subject"
  >
    <span>
      Subject
    </span>
    <i
      className="icon-arrow"
      onClick={[Function]}
    />
  </div>
  <div
    className="content-item"
  >
    &lt;p&gt;Content&lt;/p&gt;
  </div>
</div>
`;

exports[`CollapsibleRow  render isHtml = true - component render correctly - show content with Html format 1`] = `
<div
  className="collapse-item collapse"
>
  <div
    className="subject-item"
    title="Subject"
  >
    <span>
      Subject
    </span>
    <i
      className="icon-arrow"
      onClick={[Function]}
    />
  </div>
  <div
    className="content-item"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<p>Content</p>",
      }
    }
  />
</div>
`;
